geoSegment.js

```diff
@@ -0,0 +1,353 @@
+    await fs.writeJson(sessionSeriesFilePath, sessionSeriesData);
+    SESSION_SERIES_FILE_PATHS.push(sessionSeriesFilePath);
+
+    // Write ScheduledSessions to file
+    for (const generatedScheduledSession of scheduledSessionsForEveryEventSchedule) {
+      await linkScheduledSessionAndSessionSeriesAndWrite(generatedScheduledSession, sessionSeriesData);
+    }
+  });
+}
+
+/**
+ * @param {string} firehoseBaseUrl
+ * @param {string} firehoseApiKey
+ * @param {Segment[]} segments
+ */
+async function downloadSessionSeries(firehoseBaseUrl, firehoseApiKey, segments) {
+  console.log('downloadAndWriteSessionSeries() - starting');
+  // Get SessionSeries from Firehose
+  let nextUrl = `${firehoseBaseUrl}session-series`;
+  let backoffTimeInSeconds = RETRY_BACKOFF_MIN;
+  while (true) {
+    let nextNextUrl;
+    let items;
+    try {
+      ({ nextNextUrl, items } = await getFirehosePage(nextUrl, firehoseApiKey));
+    } catch (error) {
+      if (error instanceof Http404Error) {
+        console.error(`ERROR: URL ("${error.url}") not found. Please check the value for <rpde-endpoint>.`);
+        process.exit(1);
+      }
+      const loggableError = (error && error.isAxiosError)
+        ? loggableAxiosError(error)
+        : error;
+      console.warn(`WARN: Retrying [page: "${nextUrl}"] due to error:`, loggableError);
+      if (backoffTimeInSeconds > RETRY_BACKOFF_MAX) {
+        console.error('ERROR: Cannot download Firehose pages, backoff limit reached, terminating script');
+        process.exit(1);
+      }
+      await wait(backoffTimeInSeconds * 1000);
+      backoffTimeInSeconds *= RETRY_BACKOFF_EXPONENTIATION_RATE;
+      continue;
+    }
+
+    // Process SessionSeries items
+    await processSessionSeriesItems(items, segments);
+
+    // Have we reached the end of the feed?
+    if (nextNextUrl === nextUrl) {
+      break;
+    }
+
+    // If we haven't reached the end of the feed, carry on
+    nextUrl = nextNextUrl;
+    backoffTimeInSeconds = RETRY_BACKOFF_MIN;
+    await wait(MIN_REQUEST_DELAY_SECONDS * 1000);
+  }
+}
+
+/**
+ * @param {ScheduledSessionItem[]} items
+ */
+async function processScheduledSessionItems(items) {
+  items.forEach(async (scheduledSessionItem) => {
+    // Filter deleted ScheduledSessions
+    if (scheduledSessionItem.state === 'deleted') {
+      return;
+    }
+
+    // Filter ScheduledSessions in the past
+    if (!scheduledSessionItem.data.startDate || moment(scheduledSessionItem.data.startDate).isBefore(moment())) {
+      return;
+    }
+
+    // Does ScheduledSession have downloaded superEvent?
+    const scheduledSessionSuperEventId = scheduledSessionItem.data.superEvent;
+    const sessionSeriesFilePath = getSessionSeriesFilePath(scheduledSessionSuperEventId);
+    if (!SESSION_SERIES_FILE_PATHS.includes(sessionSeriesFilePath)) {
+      return;
+    }
+
+    // Get SessionSeries
+    const sessionSeries = await fs.readJson(sessionSeriesFilePath);
```

Comment:
Is there an error condition here where the SessionSeries is not found (e.g. there's a ScheduledSession with no matching SessionSeries in the data for some reason)?

Reply:
Yep corrected